models/src/node/drain.rs

```diff
@@ -0,0 +1,364 @@
+
+// After evictions are created, we wait for the Pods to be deleted by Kubernetes.
+// These constants define the poll interval for checking the Pods, as well as the max amount of time to wait.
+const DELETION_CHECK_INTERVAL: Duration = Duration::from_secs(5);
+// `kubectl drain` by default will wait "forever" for an eviction to complete. We follow suit.
+const DELETION_TIMEOUT: Duration = Duration::from_secs(u64::MAX);
+
+// Some errors while attempting evictions result in retries with exponential backoff.
+// These values configure how long to delay between tries.
+// We should be tenacious in attempting retries, as some workloads are sensitive to being suddenly interrupted.
+const RETRY_BASE_DELAY: Duration = Duration::from_millis(100);
+const RETRY_MAX_DELAY: Duration = Duration::from_secs(20);
+const NUM_RETRIES: usize = 10;
+
+/// Wrapper struct to provide retry configurations for evictions.
+struct RetryStrategy {}
+impl RetryStrategy {
+    fn retry_strategy() -> impl Iterator<Item = Duration> {
+        ExponentialBackoff::from_millis(RETRY_BASE_DELAY.as_millis() as u64)
+            .max_delay(RETRY_MAX_DELAY)
+            .map(jitter)
+            .take(NUM_RETRIES)
+    }
+}
+
+impl tokio_retry::Condition<error::EvictionError> for RetryStrategy {
+    fn should_retry(&mut self, error: &error::EvictionError) -> bool {
+        error.should_retry()
+    }
+}
+
+/// Drains a node of all pods.
+///
+/// The Kubernetes API does not provide an implementation of drain. You must use Pod deletion or the Eviction API
+/// to remove all Pods from a given Node. We opt to use Evictions in order to respect Pod Disruption Budgets.
+///
+/// The implementation of `kubectl drain` can be used as inspiration. While this implementation will halt under
+/// certain special conditions, we have slightly different behavior:
+///
+/// Kubectl by default will not evict nodes under some criteria without further instruction.
+/// By default, we ignore:
+/// * DaemonSet Pods - The DaemonSet controller will not respect node cordons, so we don't battle it.
+/// * Mirror Pods - These are static and cannot be controlled.
+///
+/// Otherwise, we evict pods that kubectl gives special care:
+/// - Pods with local storage that will be deleted when drained (emptyDir).
+/// - Unreplicated pods (Pods without a controller.)
+///
+/// PodDisruptionBudgets can be used to protect these workloads from being unduly interrupted.
+#[instrument(skip(k8s_client), err)]
+pub(crate) async fn drain_node(
+    k8s_client: &kube::Client,
+    node_name: &str,
+) -> Result<(), error::DrainError> {
+    let target_pods = find_target_pods(k8s_client, node_name).await?;
+
+    // Perform the eviction for each pod simultaneously.
+    stream::iter(target_pods)
+        .for_each(move |pod| {
+            let k8s_client = k8s_client.clone();
+            let pod = pod.clone();
+            async move {
+                if evict_pod(&k8s_client, &pod).await.is_ok() {
```

Comment:
I am a little bit confused here. 
I understand if the code is like below
```
async move {
    evict_pod(...).await?;
}
```
Then the `.await` in line 93 should catch the Err(if evict has Err) and return the Err.

But this code seems only use `is_ok()` to check evict_pod, which only returns a bool value. So what will happen if evict_pod is not ok? My understand is it will silently skip the deletion and do nothing in the stream without sending out Err.

Reply:
That's correct! Basically, if we fail to issue an eviction for a Pod, then we don't have any guarantee that the kubernetes API is deleting the Pod, so we don't wait. Because we use `#[instrument(err)]` in `evict_pod()` we will get a log message indicating that an error occurred while attempting the eviction. We don't need to propagate the error up any further, so we just drop it.

Reply:
I'll add some comments about how errors are handled to make it clearer.